phoenix-core/src/main/java/org/apache/phoenix/iterate/DeferredByteBufferSegmentQueue.java

```diff
@@ -0,0 +1,123 @@
+package org.apache.phoenix.iterate;
+
+import org.apache.commons.io.output.DeferredFileOutputStream;
+import org.apache.phoenix.memory.MemoryManager;
+import org.apache.phoenix.memory.MemoryManager.MemoryChunk;
+
+import java.io.*;
+import java.util.*;
+
+public abstract class DeferredByteBufferSegmentQueue<T> extends BufferSegmentQueue<T> {
+
+    final MemoryChunk chunk;
+
+    public DeferredByteBufferSegmentQueue(int index, int thresholdBytes,
+                                          boolean hasMaxQueueSize, MemoryManager memoryManager) {
+        super(index, thresholdBytes, hasMaxQueueSize);
+        chunk = memoryManager.allocate(thresholdBytes);
+    }
+
+    abstract protected void writeToBuffer(OutputStream outputStream, T e);
+    abstract protected T readFromBuffer(DataInput dataInput);
```

Comment:
I know they are the existing method names, but they might be even clearer if changed to "writeRecord(ToBuffer)" and "readRecord(FromBuffer)"?